src/client/content_cao.cpp

```diff
@@ -443,47 +444,100 @@ scene::IAnimatedMeshSceneNode* GenericCAO::getAnimatedMeshSceneNode()
 
 void GenericCAO::setChildrenVisible(bool toset)
 {
-	for (u16 cao_id : m_children) {
+	for (u16 cao_id : m_attachment_child_ids) {
 		GenericCAO *obj = m_env->getGenericCAO(cao_id);
 		if (obj) {
 			obj->setVisible(toset);
 		}
 	}
 }
 
-void GenericCAO::setAttachments()
+void GenericCAO::setAttachment(int parent_id, const std::string &bone, v3f position, v3f rotation)
 {
+	int old_parent = m_attachment_parent_id;
+	m_attachment_parent_id = parent_id;
+	m_attachment_bone = bone;
+	m_attachment_position = position;
+	m_attachment_rotation = rotation;
+
+	ClientActiveObject *parent = m_env->getActiveObject(parent_id);
+
+	// localplayer itself can't be attached to localplayer
+	if (!m_is_local_player) {
+		// Objects attached to the local player should be hidden by default
+		m_attached_to_local = parent && parent->isLocalPlayer();
+		m_is_visible = !m_attached_to_local;
+	}
+
+	if (parent_id != old_parent) {
+		if (auto *o = m_env->getActiveObject(old_parent))
+			o->removeAttachmentChild(m_id);
+		if (parent)
+			parent->addAttachmentChild(m_id);
+	}
+
 	updateAttachments();
 }
 
+void GenericCAO::getAttachment(int *parent_id, std::string *bone, v3f *position,
+	v3f *rotation)
+{
+	*parent_id = m_attachment_parent_id;
+	*bone = m_attachment_bone;
+	*position = m_attachment_position;
+	*rotation = m_attachment_rotation;
+}
+
+void GenericCAO::clearChildAttachments()
+{
+	// Cannot use for-loop here: setAttachment() modifies 'm_attachment_child_ids'!
+	while (!m_attachment_child_ids.empty()) {
+		int child_id = *m_attachment_child_ids.begin();
+
+		if (ClientActiveObject *child = m_env->getActiveObject(child_id))
+			child->setAttachment(0, "", v3f(0, 0, 0), v3f(0, 0, 0));
+
+		removeAttachmentChild(child_id);
+	}
+}
+
+void GenericCAO::clearParentAttachment()
+{
+	if (m_attachment_parent_id)
+		setAttachment(0, "", m_attachment_position, m_attachment_rotation);
+	else
+		setAttachment(0, "", v3f(0, 0, 0), v3f(0, 0, 0));
+}
+
+void GenericCAO::addAttachmentChild(int child_id)
+{
+	m_attachment_child_ids.insert(child_id);
+}
+
+void GenericCAO::removeAttachmentChild(int child_id)
+{
+	m_attachment_child_ids.erase(child_id);
+}
+
 ClientActiveObject* GenericCAO::getParent() const
 {
 	ClientActiveObject *obj = NULL;
 
-	u16 attached_id = m_env->attachement_parent_ids[getId()];
+	if (m_attachment_parent_id)
+		obj = m_env->getActiveObject(m_attachment_parent_id);
 
-	if ((attached_id != 0) &&
-			(attached_id != getId())) {
-		obj = m_env->getActiveObject(attached_id);
-	}
 	return obj;
```

Comment:
Why not like this?
```cpp
	return m_attachment_parent_id ? m_env->getActiveObject(m_attachment_parent_id) : nullptr;
```

Reply:
Just code style and readability, probably. Complex long conditions are less readable when using ternary operators.

Reply:
If I'm not mistaken, the extra pair of parenthesis around the individual conditions aren't required.